packages/compartment-mapper/src/policy.js

```diff
@@ -0,0 +1,305 @@
+  if (!localPolicy) {
+    return globals;
+  }
+  const list = getGlobalsList(localPolicy);
+  return copyGlobals(globals, list);
+};
+
+/**
+ * Throws if importing of the specifier is not allowed by the policy
+ *
+ * @param {string} specifier
+ * @param {Object} compartmentDescriptor
+ * @param {Object} [info]
+ */
+export const gatekeepModuleAccess = (
+  specifier,
+  compartmentDescriptor,
+  info,
+) => {
+  const { policy, modules } = compartmentDescriptor;
+  if (!policy) {
+    return;
+  }
+
+  if (!info.exit) {
+    if (!modules[specifier]) {
+      throw Error(
+        `Importing '${specifier}' was not allowed by policy packages:${JSON.stringify(
+          policy.packages,
+        )}`,
+      );
+    }
+    return;
+  }
+
+  if (!policy.builtins || !policy.builtins[specifier]) {
+    throw Error(
+      `Importing '${specifier}' was not allowed by policy 'builtins':${JSON.stringify(
+        policy.builtins,
+      )}`,
+    );
+  }
+};
+
+function attenuateModule({ attenuators, name, params, originalModule }) {
+  if (!attenuators) {
+    // TODO: figure out what to pass here or where to move this check
+    throw Error(
+      `Attenuation '${name}' in policy doesn't have a corresponding implementation.`,
+    );
+  }
+  const attenuationCompartment = new Compartment(
+    {},
+    {},
+    {
+      name,
+      resolveHook: moduleSpecifier => moduleSpecifier,
+      importHook: async () => {
+        const {
+          namespace: { attenuate },
+        } = await attenuators(name);
+        const ns = await attenuate(params, originalModule);
+        const staticModuleRecord = freeze({
+          imports: [],
+          exports: keys(ns),
+          execute: moduleExports => {
+            assign(moduleExports, ns);
+          },
+        });
+        return staticModuleRecord;
+      },
+    },
+  );
+  return attenuationCompartment.module('.');
+}
+
+/**
+ * Throws if importing of the specifier is not allowed by the policy
+ *
+ * @param {string} specifier - exit module name
+ * @param {Object} originalModule - reference to the exit module
+ * @param {Object} policy - local compartment policy
+ * @param {Object} attenuators - a key-value where attenuations can be found
+ */
+export const attenuateModuleHook = (
+  specifier,
+  originalModule,
+  policy,
+  attenuators,
+) => {
+  if (policy && (!policy.builtins || !policy.builtins[specifier])) {
+    throw Error(
+      `Attenuation failed '${specifier}' was not in policy builtins:${stringify(
+        policy.builtins,
+      )}`,
+    );
+  }
+  if (!policy || policy.builtins[specifier] === true) {
+    return originalModule;
+  }
+
+  return attenuateModule({
+    attenuators,
+    name: policy.builtins[specifier].attenuate,
+    params: policy.builtins[specifier].params,
+    originalModule,
+  });
+};
+
+const padDiagnosis = text => ` (${text})`;
+/**
+ * Provide dignostic information for a missing compartment error
+ *
+ * @param {Object}  params
```

Comment:
I don’t know why we are heading in this stylistic direction, but spelling `object` in the lesser case is the prevailing trend. I have favored `args` and `opts` depending on whether this is a bag of required or optional values.
```suggestion
 * @param {object}  args
```

Reply:
Typescript enthusiasts told me to use Object not object. I no longer remember why tho. Cargo culting it ever since.
It's also 3x more popular in this codebase. We can make it consistent in a separate PR